File: src/com/xw/vvtts/utils/VoiceConfig.java
package com.xw.vvtts.utils;

import android.content.Context;
import android.content.SharedPreferences;

public class VoiceConfig {
    private static final String PREFS = "vvtts_prefs";
    public static final String KEY_VOICE = "voice";
    public static final String KEY_RATE = "rate";
    public static final String KEY_PITCH = "pitch";
    public static final String KEY_VOLUME = "volume";
    public static final String KEY_AUTO_DETECT = "auto_detect";

    /** 支持的语言定义（苹果 Kona 全语言表） */
    public static class Lang {
        public final String code;    // BCP-47
        public final String name;    // 中文名
        public final int konaDialect;// 苹果 Kona dialect 编号
        public final long eciDialect;// ECI dialect（部分为推测值，接入库时实测修正）
        public Lang(String code, String name, int konaDialect, long eciDialect) {
            this.code = code; this.name = name;
            this.konaDialect = konaDialect; this.eciDialect = eciDialect;
        }
    }

    // 已知锚点：en-US kona0=0x10000，zh-CN kona12=0x60000（广荣实测）
    // 其余为对照 Eloquence 官方 dialect 家族的推测值，接入对应语言库时逐一验证
    public static final Lang[] LANGS = {
        new Lang("en-US", "英语（美式）", 0,  0x10000),
        new Lang("en-GB", "英语（英式）", 1,  0x20000),
        new Lang("es-ES", "西班牙语（西班牙）", 2, 0x30000),
        new Lang("es-MX", "西班牙语（墨西哥）", 3, 0x30000),
        new Lang("fr-FR", "法语（法国）", 4,  0x0C0000),
        new Lang("fr-CA", "法语（加拿大）", 5, 0x0C0000),
        new Lang("de-DE", "德语", 6,  0x70000),
        new Lang("it-IT", "意大利语", 7, 0x100000),
        new Lang("pt-BR", "葡萄牙语（巴西）", 8, 0x160000),
        new Lang("fi-FI", "芬兰语", 9, 0x0B0000),
        new Lang("ja-JP", "日语", 10, 0x110000),
        new Lang("ko-KR", "韩语", 11, 0x120000),
        new Lang("zh-CN", "中文（普通话）", 12, 0x60000),
        new Lang("zh-TW", "中文（台湾）", 13, 0x61000),
    };

    // CF 语言库代码（Code Factory 10 语言）与 BCP-47 对应
    public static String cfCodeFor(String bcp47) {
        if (bcp47 == null) return null;
        switch (bcp47) {
            case "de-DE": return "deu";
            case "en-US": return "enu";
            case "en-GB": return "eng";
            case "es-ES": return "esn";  // Castilian Spanish
            case "es-MX": return "esm";  // Latin American / Mexican Spanish
            case "fr-FR": return "fra";
            case "fr-CA": return "frc";
            case "it-IT": return "ita";
            case "pt-BR": return "ptb";
            case "fi-FI": return "fin";
            default: return null;      // zh/ja/ko 无 CF 库 → 广荣链路
        }
    }

    public static Lang findLang(String code) {
        if (code == null) return LANGS[12];
        for (Lang l : LANGS) if (l.code.equals(code)) return l;
        return LANGS[12];
    }

    public static Lang findLangByKona(int kona) {
        for (Lang l : LANGS) if (l.konaDialect == kona) return l;
        return LANGS[12];
    }

    private final SharedPreferences prefs;

    public VoiceConfig(Context ctx) {
        prefs = ctx.getSharedPreferences(PREFS, Context.MODE_PRIVATE);
    }

    public String getVoice() { return prefs.getString(KEY_VOICE, "zh-CN"); }
    public int getRate() { return prefs.getInt(KEY_RATE, 100); }
    public int getPitch() { return prefs.getInt(KEY_PITCH, 50); }
    public int getVolume() { return prefs.getInt(KEY_VOLUME, 100); }
    public boolean isAutoDetect() { return prefs.getBoolean(KEY_AUTO_DETECT, true); }

    public void setVoice(String v) { prefs.edit().putString(KEY_VOICE, v).commit(); }
    public void setRate(int r) { prefs.edit().putInt(KEY_RATE, r).commit(); }
    public void setPitch(int p) { prefs.edit().putInt(KEY_PITCH, p).commit(); }
    public void setVolume(int v) { prefs.edit().putInt(KEY_VOLUME, v).commit(); }
    public void setAutoDetect(boolean b) { prefs.edit().putBoolean(KEY_AUTO_DETECT, b).commit(); }
}

File: src/com/xw/vvtts/utils/VoiceConfig.kt
package com.xw.vvtts.utils

import android.content.Context
import android.content.SharedPreferences

/** Voice settings (UI rate/pitch/volume + language selection).) */
class VoiceConfig(context: Context) {
    /** Supported language definitions (Apple Kona full table).)*/
    class Lang(
        /** BCP-47 */
        val code: String,
        /** Display name */
        val name: String,
        /** Apple Kona dialect number */
        val konaDialect: Int,
        /** ECI dialect (partly derived; confirmed against the linked modules) */
        val eciDialect: Long,
    )

    private val prefs: SharedPreferences = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

    val voice: String
        get() = prefs.getString(KEY_VOICE, "en-US")!!
    val rate: Int
        get() = prefs.getInt(KEY_RATE, 100)
    val pitch: Int
        get() = prefs.getInt(KEY_PITCH, 50)
    val volume: Int
        get() = prefs.getInt(KEY_VOLUME, 100)
    val isAutoDetect: Boolean
        get() = prefs.getBoolean(KEY_AUTO_DETECT, true)

    fun setVoice(v: String) { prefs.edit().putString(KEY_VOICE, v).apply() }
    fun setRate(r: Int) { prefs.edit().putInt(KEY_RATE, r).apply() }
    fun setPitch(p: Int) { prefs.edit().putInt(KEY_PITCH, p).apply() }
    fun setVolume(v: Int) { prefs.edit().putInt(KEY_VOLUME, v).apply() }
<<<<<<< HEAD
=======
    fun setDspMode(m: Int) { prefs.edit().putInt(KEY_DSP_MODE, m).apply() }
>>>>>>> 9b38c1c (utils: same-set Lingua guard, emoji scan gate, async prefs)
    fun setAutoDetect(b: Boolean) { prefs.edit().putBoolean(KEY_AUTO_DETECT, b).apply() }

    fun setPunctEnabled(b: Boolean) { prefs.edit().putBoolean(KEY_PUNCT, b).apply() }
    val punctEnabled: Boolean
        get() = prefs.getBoolean(KEY_PUNCT, false)

    /** Dictionary: newline-separated "word|spoken" lines in creation order. */
    fun dictEntries(): List<Pair<String, String>> {
        val raw = prefs.getString(KEY_DICT, "") ?: ""
        val out = ArrayList<Pair<String, String>>()
        for (line in raw.split("\n")) {
            val idx = line.indexOf('|')
            if (idx > 0 && idx < line.length - 1) {
                out.add(Pair(line.substring(0, idx).trim(), line.substring(idx + 1).trim()))
            }
        }
        return out
    }

    fun addDictEntry(word: String, spoken: String) {
        val w = word.trim()
        val s = spoken.trim()
        if (w.isEmpty() || s.isEmpty()) return
        val cur = prefs.getString(KEY_DICT, "") ?: ""
        val kept = ArrayList<String>()
        for (l in cur.split("\n")) {
            if (l.isBlank()) continue
            if (l.substringBefore('|').equals(w, ignoreCase = true)) continue
            kept.add(l)
        }
        kept.add(w + "|" + s)
        prefs.edit().putString(KEY_DICT, kept.joinToString("\n")).apply()
    }

    fun removeDictEntry(word: String) {
        val cur = prefs.getString(KEY_DICT, "") ?: ""
        val kept = ArrayList<String>()
        for (l in cur.split("\n")) {
            if (l.isBlank()) continue
            if (l.substringBefore('|').equals(word, ignoreCase = true)) continue
            kept.add(l)
        }
        prefs.edit().putString(KEY_DICT, kept.joinToString("\n")).apply()
    }

    fun clearDict() { prefs.edit().remove(KEY_DICT).apply() }
    companion object {
        private const val PREFS = "vvtts_prefs"
        const val KEY_VOICE = "voice"
        const val KEY_RATE = "rate"
        const val KEY_PITCH = "pitch"
        const val KEY_VOLUME = "volume"
        const val KEY_AUTO_DETECT = "auto_detect"
        const val KEY_PUNCT = "speak_punctuation"
        const val KEY_DICT = "user_dict"

        // eciDialect values: measured from each language module's registered constant
        // (lang/*/eci_ini_*.c *_eci_library_lang); one-to-one with build_native.sh LANGS
        // 0 = language not linked here (ko/zh-TW have no module; zh-CN IS linked via oracle)
        val LANGS = arrayOf(
            Lang("en-US", "English (US)", 0, 0x10000L),
            Lang("en-GB", "English (UK)", 1, 0x10001L),
            Lang("es-ES", "Spanish (Spain)", 2, 0x20000L),
            Lang("es-US", "Spanish (US)", -1, 0x20001L), // esus module
            Lang("es-MX", "Spanish (Mexico)", 3, 0x20002L), // esmx module
            Lang("fr-FR", "French (France)", 4, 0x30000L),
            Lang("fr-CA", "French (Canada)", 5, 0x30001L),
            Lang("de-DE", "German", 6, 0x40000L),
            Lang("it-IT", "Italian", 7, 0x50000L),
            Lang("pt-BR", "Portuguese (Brazil)", 8, 0x70000L),
            Lang("fi-FI", "Finnish", 9, 0x90000L),
            Lang("ja-JP", "Japanese", 10, 0x80000L),
            Lang("pl-PL", "Polish", -1, 0x110000L), // plpl module
            Lang("ko-KR", "Korean", 11, 0L), // not linked
            Lang("zh-CN", "Chinese (Mandarin)", 12, 0x60000L), // linked: oracle synth path
            Lang("zh-TW", "Chinese (Taiwan)", 13, 0x60001L), // not linked
        )

        /** CF library code (Code Factory 10 languages) mapped from BCP-47 */
        fun cfCodeFor(bcp47: String?): String? {
            if (bcp47 == null) return null
            return when (bcp47) {
                "de-DE" -> "deu"
                "en-US" -> "enu"
                "en-GB" -> "eng"
                "es-ES" -> "esn"  // Castilian Spanish
                "es-MX" -> "esm"  // Latin American / Mexican Spanish
                "fr-FR" -> "fra"
                "fr-CA" -> "frc"
                "it-IT" -> "ita"
                "pt-BR" -> "ptb"
                "fi-FI" -> "fin"
                else -> null      // zh/ja/ko have no CF library → the openevv chain handles them
            }
        }

        fun findLang(code: String?): Lang {
            if (code == null) return LANGS[0]// en-US
            for (l in LANGS) if (l.code == code) return l
            return LANGS[0] // en-US
        }
    }
}